chapter_02B_abstractions.asciidoc

```diff
@@ -522,107 +522,57 @@ Our tests now act directly on the `determine_actions()` function:
-    src_hashes = reader(src_folder)  #<2>
-    dst_hashes = reader(dst_folder)  #<2>
+def synchronise_dirs(reader, filesystem, source_root, dest_root): #<1>
+
+    source_hashes = reader(source_root) #<2>
+    dest_hashes = reader(dest_root)
 
     for sha, filename in src_hashes.items():
         if sha not in dst_hashes:
-            sourcepath = src_folder / filename
-            destpath = dst_folder / filename
-            apply_func('COPY', sourcepath, destpath)  #<3>
+            sourcepath = source_root / filename
+            destpath = dest_root / filename
+            filesystem.copy(destpath, sourcepath) #<3>
 
         elif dst_hashes[sha] != filename:
-            olddestpath = dst_folder / dst_hashes[sha]
-            newdestpath = dst_folder / filename
-            apply_func('MOVE', olddestpath, newdestpath)  #<3>
+            olddestpath = dest_root / dst_hashes[sha]
+            newdestpath = dest_root / filename
+            filesystem.move(oldestpath, newdestpath)
 
     for sha, filename in dst_hashes.items():
         if sha not in src_hashes:
-            apply_func('DELETE', dst_folder / filename)  #<3>
-
+            filesystem.del(dest_root/filename)
 ----
 ====
 
-// TODO (DS): Why not just call reader `read_paths_and_hashes`, like in the
-// FCIS example?
-
-//NICE-TO-HAVE: test this listing
-
-<1> Our top-level function now exposes two new dependencies, a `reader` and an
-    `apply_func`
+<1> Our top-level function now exposes two new dependencies, a `reader` and a
+    `filesystem`
 
-<2> We invoke the `reader` to produce our dict-abstraction of the filesystems
+<2> We invoke the `reader` to produce our files dict.
 
-<3> And we invoke the `apply_func` using our action-abstraction for the actions
-    we want to apply.
+<3> And we invoke the `filesystem` to apply the changes we detect.
 
 TIP: Notice that, although we're using dependency injection, there was no need
     to define an abstract base class or any kind of explicit interface.  In the
     book we often show ABCs because we hope they help to understand what the
     abstraction is, but they're not necessary.  Python's dynamic nature means
     we can always rely on duck typing.
-
-////
-TODO (DS)
-Really, the only difference between the DI approach shown here and the FCIS is
-that in DI, the orchestration logic can be tested against abstractions too (at
-the cost of some extra indirection).
-
-It feels to me that this DI approach is building on the FCIS, rather than
-demonstrating the 'pure' version of DI. In real life, it's nice to combine
-them, but perhaps it would be a clearer comparison if you showed a purer
-version of each.
-
-I also think you might need to address how the patterns you propose in this
-book relate to the alternative approaches. To me, it seems the core patterns
-are much more DI than FCIS. But you could also point out that the approaches
-can be combined - it's not either/or.
-////
-
```

Comment:
I'd say that the big difference between the DI and FCIS snippets is that the the DI sample has side effects interleaved with the processing logic.  The FCIS snippet is clearly separable between 3 steps: 1. input side-effects when reading the filesystem and hashing, 2. pure data processing (aka the core), then 3.  output side-effects when writing to the filesystem.

You might also be able to say something about the Filesystem type encapsulating a set of specific responsibilities that form a role, which does not exist in the FCIS example.